Mathlib/Algebra/Polynomial/Factors.lean

```diff
@@ -0,0 +1,204 @@
+
+## Main definitions
+
+* `Polynomial.Factors f`: A predicate on a polynomial `f` saying that `f` is a product of
+  constant and monic linear polynomials.
+
+-/
+
+variable {R : Type*}
+
+namespace Polynomial
+
+section Semiring
+
+variable [Semiring R]
+
+/-- A polynomial factors if a product of constant and monic linear polynomials. -/
+def Factors (f : R[X]) : Prop := f ∈ Submonoid.closure ({C a | a : R} ∪ {X + C a | a : R})
+
+theorem factors_C (a : R) : Factors (C a) :=
+  Submonoid.mem_closure_of_mem (Set.mem_union_left _ ⟨a, rfl⟩)
+
+theorem factors_zero : Factors (0 : R[X]) := by
+  simpa using factors_C (0 : R)
+
+theorem factors_one : Factors (1 : R[X]) := by
+  simpa using factors_C (1 : R)
+
+theorem factors_X_add_C (a : R) : Factors (X + C a) :=
+  Submonoid.mem_closure_of_mem (Set.mem_union_right _ ⟨a, rfl⟩)
+
+theorem factors_X : Factors (X : R[X]) := by
+  simpa using factors_X_add_C (0 : R)
+
+protected theorem Factors.mul {f g : R[X]} (hf : Factors f) (hg : Factors g) :
+    Factors (f * g) :=
+  mul_mem hf hg
+
+theorem Factors.C_mul {f : R[X]} (hf : Factors f) (a : R) : Factors (C a * f) :=
+  (factors_C a).mul hf
+
+theorem Factors.list_prod {l : List R[X]} (h : ∀ f ∈ l, Factors f) : Factors l.prod := by
+  induction l with
+  | nil => exact factors_one
+  | cons f l ih =>
+    exact (h f List.mem_cons_self).mul (ih (h · ∘ List.mem_cons_of_mem f))
+
+protected theorem Factors.pow {f : R[X]} (hf : Factors f) (n : ℕ) : Factors (f ^ n) := by
+  rw [← List.prod_replicate]
+  exact Factors.list_prod (by simp [hf])
+
+theorem factors_X_pow (n : ℕ) : Factors (X ^ n : R[X]) :=
+  factors_X.pow n
+
+theorem factors_C_mul_X_pow (a : R) (n : ℕ) : Factors (C a * X ^ n) :=
+  (factors_X_pow n).C_mul a
+
+theorem factors_monomial (n : ℕ) (a : R) : Factors (monomial n a) := by
+  simp [← C_mul_X_pow_eq_monomial, factors_C_mul_X_pow]
+
+protected theorem Factors.map {f : R[X]} (hf : Factors f) {S : Type*} [Semiring S] (i : R →+* S) :
+    Factors (map i f) := by
+  have h := Submonoid.mem_map_of_mem (mapRingHom i) hf
+  rw [MonoidHom.map_mclosure, Set.image_union] at h
+  refine Submonoid.closure_mono (Set.union_subset_union ?_ ?_) h <;>
+    rintro - ⟨-, ⟨a, rfl⟩, rfl⟩ <;> exact ⟨i a, by simp⟩
+
+theorem factors_of_isUnit [NoZeroDivisors R] {f : R[X]} (hf : IsUnit f) : Factors f :=
+  (isUnit_iff.mp hf).choose_spec.2 ▸ factors_C _
+
+end Semiring
+
+section CommSemiring
+
+variable [CommSemiring R]
+
+theorem Factors.multiset_prod {m : Multiset R[X]} (hm : ∀ f ∈ m, Factors f) : Factors m.prod := by
+  rw [← Multiset.prod_toList]
+  exact Factors.list_prod (by simpa)
+
+protected theorem Factors.prod {ι : Type*} {f : ι → R[X]} {s : Finset ι}
+    (h : ∀ i ∈ s, Factors (f i)) : Factors (∏ i ∈ s, f i) := by
+  rw [Finset.prod_eq_multiset_prod]
+  exact Factors.multiset_prod (by simpa)
+
+theorem factors_iff_exists_multiset_add {f : R[X]} :
+    Factors f ↔ ∃ m : Multiset R, f = C f.leadingCoeff * (m.map (X + C ·)).prod := by
```

Comment:
```suggestion
theorem factors_iff_exists_multiset {f : R[X]} :
    Factors f ↔ ∃ m : Multiset R, f = C f.leadingCoeff * (m.map (X + C ·)).prod := by
```

Reply:
This would clash with `factors_iff_exists_multiset` below, but I've changed this to a prime-d version with a docstring.